### PR TITLE
Slechta/issue 262

### PR DESCRIPTION
**Pull request by slechta**
Fixes a bug where empty partitions and/or leader in kafka were not considered as error.

### CODE REVIEW

File: euphoria-kafka/src/main/java/cz/seznam/euphoria/kafka/KafkaSource.java [hunk 1]
```diff
@@ -219,17 +219,25 @@ public KafkaSource(
...
+          throw new IllegalStateException("Leader not available");
+        }
+
+        return new KafkaPartition(
+          brokerList, topicId, p.partition(),
+          config,
+          offs.getOrDefault(p.partition(), defaultOffsetTimestamp),
+          stopAtStamp);
+        })
+      .collect(Collectors.toList());
```
Comment (je-ik) [inline euphoria-kafka/src/main/java/cz/seznam/euphoria/kafka/KafkaSource.java]: This should be probably indented. Thanks! :-)

**Comment by slechta** [comment]
In order to make the test I would need to mock static method KafkaUtils.newConsumer(). Mockito does not support mocking static methods, therefore I would need adding PowerMock to test dependecies.

**Comment by dmvk** [comment]
Shouldn't we rather replace it with a factory interface, than introducing new dependency?

**Comment by je-ik** [comment]
I don't think it is necessary. Mocking static methods sounds weird to me, I'd suggest the following small refactor:
```java
    @Override
    public UnboundedReader<Pair<byte[], byte[]>, Long> openReader() throws IOException {
      Consumer<byte[], byte[]> c = newConsumer(brokerList, config);
      TopicPartition tp = new TopicPartition(topicId, partition);
      ArrayList<TopicPartition> partitionList = Lists.newArrayList(tp);
      c.assign(partitionList);
      if (startOffset > 0) {
        c.seek(tp, startOffset);
      } else if (startOffset == 0) {
        c.seekToBeginning(partitionList);
      }
      return new ConsumerReader(c, tp, stopReadingAtStamp);
    }

    @VisibleForTesting
    KafkaConsumer<byte[], byte[]> newConsumer(String brokerList, Settings config) {
      return KafkaUtils.newConsumer(brokerList, null, config);
    }
```
which will allow you to override the broker creation in test.

**Comment by slechta** [comment]
That is not sufficient
Mockito cannot mock/spy because :
 - final class

I will move the inner classes out, because it is already a long class. 
(Anyway I am not a big fan of mocking)

File: euphoria-kafka/src/main/java/cz/seznam/euphoria/kafka/KafkaSource.java [hunk 2]
```diff
@@ -206,7 +207,7 @@ public KafkaSource(
             stopReadingAtStamp);
       }
     }
-    try (Consumer<?, ?> c = KafkaUtils.newConsumer(
+    try (Consumer<?, ?> c = newConsumer(
```
Comment (dmvk) [inline euphoria-kafka/src/main/java/cz/seznam/euphoria/kafka/KafkaSource.java]: shouldnt there be `Consumer<byte[], byte[]>`?

**Comment by je-ik** [comment]
Cool, thanks! :+1: Merging.

**Pull request merged**